File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
            if(head->next==NULL){
            if(n==1){
                return NULL;
            }
        }
        int cnt=0;
        ListNode*temp=head;
        while(temp){
            cnt++;
            temp=temp->next;
        }
        ListNode*prev=NULL;
        temp=head;
        // cnt++;
        int revcnt=cnt+1;
        while(revcnt>0){
            revcnt--;
            if(revcnt==n){
                if(revcnt==cnt){
                   head=head->next;
                   delete(temp);
                   return head;
                }
                // else if(temp->next==NULL){
                //     prev->next=NULL;
                //     delete(temp);
                //     break;
                // }
                else{
                prev->next=temp->next;
                delete(temp);
                break;
                }
               
            }
            prev=temp;
            temp=temp->next;
        }
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int x) {
        //calculating the lenght of the node
        int len=0;
        ListNode temp=head;
        while(temp!=null){
            //jeheto count ta 0 theke start hocche tai age oi node er jonno length take barabo tarpor pointer thake next node e nie jabo ba onno kichu check korbo
            len++;
            temp=temp.next;
        }
        //ebar ami (n-x+1)th node er ager node e pointer ta nie jabo jeheto ami (n-x+1)th node take delete korbo
        //so ami pointer take (n-x+1)-1=(n-x) th node e nie jabo..mne current jei node ta delete hobe tar previous node e
        //kintu ekhane jodi head node take delete korte bole mne x=length hoi to alada bhabe handle korte hobe karon amra sekhetre 0th node take khujbo kintu 0th node paboi na karon node er numbering 1 theke start hobe
        if(x==len){
            //deleting head node
            head=head.next;
        }
        else{
        //to count the number of node traversed
        int cnt=0;
        temp=head;
        while(temp!=null){
            cnt++;
            if(cnt==(len-x)){
                break;
            }
            temp=temp.next;
        }
        //delete korar process previous node er next e , tar next node er next ke assign korlam
        temp.next=temp.next.next;
    }
        return head;
    }
    }